llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,638 @@
+        "grpc error (" + toStringRef(Status.error_code()) +
+            "): " + Status.error_message());
+  }
+
+  Error createGRPCResponseError(
+      const compilation_cache_service::cas::v1::ResponseError &RE) const {
+    return createStringError(std::make_error_code(std::errc::bad_message),
+                             "grpc error: " + RE.description());
+  }
+
+  Error createGRPCResponseError(
+      const compilation_cache_service::keyvalue::v1::ResponseError &RE) const {
+    return createStringError(std::make_error_code(std::errc::bad_message),
+                             "grpc error: " + RE.description());
+  }
+
+  // Stub for compile_cache_service.
+  std::unique_ptr<CASDBService::Stub> CASStub;
+  std::unique_ptr<KeyValueDB::Stub> KVStub;
+
+  // Index to manage the remote CAS.
+  InMemoryIndexT Index;
+  // Allocator for CAS content.
+  ThreadSafeAllocator<BumpPtrAllocator> Alloc;
+
+  // Store the size of the hash.
+  size_t ServiceHashSize;
+};
+
+} // namespace
+
+GRPCRelayCAS::GRPCRelayCAS(StringRef Path, Error &Err) {
+  ErrorAsOutParameter ErrAsOutParam(&Err);
+  std::string Address("unix:");
+  Address += Path;
+  ChannelArguments Args;
+  Args.SetMaxReceiveMessageSize(INT_MAX);
+  Args.SetMaxReceiveMessageSize(INT_MAX);
+  const auto Channel = grpc::CreateCustomChannel(
+      std::move(Address), grpc::InsecureChannelCredentials(), Args);
+  CASStub = CASDBService::NewStub(Channel);
+  KVStub = KeyValueDB::NewStub(std::move(Channel));
+
+  // Send a put request to the CAS to determine hash size.
+  CASPutRequest Request;
+  ClientContext Context;
+  CASPutResponse Response;
+  Request.mutable_data()->CopyFrom(createCASData(None, ""));
+  grpc::Status Status = CASStub->Put(&Context, Request, &Response);
+  if (!Status.ok()) {
+    Err = createGRPCStatusError(Status);
+    return;
+  }
+  if (Response.has_error()) {
+    Err = createGRPCResponseError(Response.error());
+    return;
+  }
+  ServiceHashSize = Response.cas_id().id().size();
+  if (ServiceHashSize > 80)
+    Err = createStringError(std::make_error_code(std::errc::message_size),
+                            "Hash from CASService is too long");
+}
+
+StringRef GRPCRelayCAS::getHashSchemaIdentifier() const {
+  return "grpc::cas::service::v1";
+}
+
+ArrayRef<uint8_t> GRPCRelayCAS::getHashImpl(const CASID &ID) const {
+  ArrayRef<uint8_t> ExtendedHash(extractIndexFromID(ID).Hash);
+  return ExtendedHash.take_front(ServiceHashSize);
+}
+
+// FIXME: this is a partial implementation of current CAS implementation.
+// Should move this into GRPC protocol.
+static StringRef getCASIDPrefix() { return "0~"; }
+
+static std::string decodeBase64(std::string const &encoded_string) {
+  size_t in_len = encoded_string.size();
+  size_t i = 0;
+  size_t j = 0;
+  int in_ = 0;
+  unsigned char char_array_4[4], char_array_3[3];
+  std::string ret;
+
+  auto is_base64 = [](unsigned char c) {
+    return (isalnum(c) || (c == '+') || (c == '/'));
```

Comment:
* `isalnum` can be very expensive in non-C locales. Maybe a custom range check would be better?
* If we're doing a custom decoding anyway, maybe switching to Base64URL would provide broader compatibility? https://www.rfc-editor.org/rfc/rfc4648#section-5

Reply:
The fixup above says this function (encoding and decoding of the hash) should be provided by CASService (through gRPC).

I implement the our hash encoding here just so that it is easier to debug. Technically, I can just hexdump the hash bytes and it would be the same.